### PR TITLE
driver: i2c: cc13xx_cc26xx: to support i2c scan

### LINKED ISSUE #38645: cc13xx and cc26xx  serial boards `i2c scan` not work
**Describe the bug**
I believe that's a common issue. because `i2c scan` function is merged recently. 
https://github.com/zephyrproject-rtos/zephyr/pull/21797

**To Reproduce**
1. mkdir build; cd build
2. cmake -DBOARD=cc1352r1_launchxl
3. make

**Expected behavior**
`i2c can ` work like this.
```shell
uart:~$ i2c scan I2C_2
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:             -- -- -- -- -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- 1e --
20: -- -- -- -- -- -- -- -- -- 29 -- -- -- 2d -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: -- -- -- -- -- -- 56 -- -- -- -- -- -- 5d -- 5f
60: -- -- -- -- -- -- -- -- -- -- 6a -- -- -- -- --
70: -- -- -- -- -- -- -- --
7 devices found on I2C_2
```
**Impact**
The impact is minor, it's just not functional.

**Logs and console output**
No such logs

**Environment (please complete the following information):**
common issue, so ignore the environment

**Additional context**
nothing

### PR DESCRIPTION
**Pull request by Pillar1989**
Sending address without data, make `i2c scan I2C_0` work.

Signed-off-by: Baozhu Zuo <zuobaozhu@gmail.com>

### CODE REVIEW

**Review by cfriedt — APPROVED** [review]
@Pillar1989 - please file an issue and link it to this PR. Describe the bug in more detail.

I know what this is for, but other reviewers might not.

It's mainly just because we are in our stabilization period right before release.

**Comment by Pillar1989** [comment]
thanks @cfriedt  guide me , that's the issue.  https://github.com/zephyrproject-rtos/zephyr/issues/38645

**Comment by cfriedt** [comment]
Fixes #38645

**Comment by cfriedt** [comment]
@Pillar1989 - just a few minor whitespace formatting issues. Please fix up so that CI passes and then it should be merged relatively quickly.

**Comment by Pillar1989** [comment]
update

**Review by devanlai — APPROVED** [review]
I've tested this with an LPSTK-CC1352R with the `i2c scan` command from the i2c shell and I can confirm that scanning works as expected.

It also would've been great to have this patch a few weeks ago when I was bringing up some new hardware and was puzzled by the i2c scan not showing anything.

File: drivers/i2c/i2c_cc13xx_cc26xx.c
Comment (stephanosio) [inline drivers/i2c/i2c_cc13xx_cc26xx.c:66]: Does this really work as expected?

As far as I can see, `I2C_MASTER_CMD_SINGLE_SEND` sends both address and data (written to `I2CMDR`); and we do not write to `I2CMDR` in this case, so we will be sending garbage data here:

image

Can you post an oscilloscope or logic analyser capture of the SCL/SDA waveforms and confirm that no data is sent here?

**Review by stephanosio — CHANGES_REQUESTED** [review]
Blocking this until we can validate this change.

**Comment by cfriedt** [comment]
@Pillar1989 - try to pull from `main` and rebase your branch. I vaguely recall that this `MAINTAINERS` issue was solved that way.

File: drivers/i2c/i2c_cc13xx_cc26xx.c
Comment (devanlai) [reply drivers/i2c/i2c_cc13xx_cc26xx.c:66]: Your trace matches what I see with a logic analyzer. It is disappointing that the peripheral doesn't seem to be able to send just the address without also pushing a data byte.

image

Comment (Pillar1989) [reply drivers/i2c/i2c_cc13xx_cc26xx.c:66]: @leadercxn let's figure out why there is append a  data.

File: drivers/i2c/i2c_cc13xx_cc26xx.c
Comment (cfriedt) [inline drivers/i2c/i2c_cc13xx_cc26xx.c:61]: I'm not 100% convinced that this statement is true, as I'm sure is the case for everyone else working with this part.

Let's ask the contributor with arguably the most experience with these parts.

@vanti - is there a way to perform an i2c scan using the simplelink I2C driverlib? If so, how do we do it?

Comment (cfriedt) [reply drivers/i2c/i2c_cc13xx_cc26xx.c:61]: For reference, a good description of the goal is here:
https://hackaday.com/2015/06/25/embed-with-elliot-i2c-bus-scanning/

and TI's API is here:
https://software-dl.ti.com/simplelink/esd/simplelink_msp432e4_sdk/2.30.00.14/docs/driverlib/msp432e4/html/group__i2c__api.html

File: drivers/i2c/i2c_cc13xx_cc26xx.c
Comment (leadercxn) [reply drivers/i2c/i2c_cc13xx_cc26xx.c:66]: I consult this document -- CC13x2, CC26x2 SimpleLink™ Wireless MCU Technical Reference Manual (Rev. D) , in chapter 25.3.5 , it introduces several models of I2C communication, but i can't see any modes that just send the address without I2C_MDR data.
In the chapter 23.5 , there are no any registers refer to how to just send the address. I try to modify the MCTRL register, but it doesn't work . so i don‘t think it's hardware I2C  supports this feature that  just send the address without I2C_MDR data.

i2c_scan

[CC13x2, CC26x2 SimpleLink™ Wireless MCU Technical Reference Manual](https://www.ti.com/lit/ug/swcu185d/swcu185d.pdf?ts=1632376017614&ref_url=https%253A%252F%252Fwww.ti.com%252Fproduct%252FCC1352P)

Comment (Pillar1989) [reply drivers/i2c/i2c_cc13xx_cc26xx.c:66]: @devanlai  @stephanosio  It seems that after some research, this additional data is justified.

Comment (stephanosio) [reply drivers/i2c/i2c_cc13xx_cc26xx.c:66]: > It seems that after some research, this additional data is justified.

@Pillar1989 Could you elaborate?

**Comment by Pillar1989** [comment]
> @Pillar1989 - try to pull from `main` and rebase your branch. I vaguely recall that this `MAINTAINERS` issue was solved that way.

let me try again

File: drivers/i2c/i2c_cc13xx_cc26xx.c
Comment (Pillar1989) [reply drivers/i2c/i2c_cc13xx_cc26xx.c:66]: In chapter 23.5, which has many Master TRANSMIT modes.  But every mode needs append data after the i2c address. 

image

**Comment by Pillar1989** [comment]
@stephanosio any more questions need to discuss ?

File: drivers/i2c/i2c_cc13xx_cc26xx.c
Comment (stephanosio) [reply drivers/i2c/i2c_cc13xx_cc26xx.c:66]: > In chapter 23.5, which has many Master TRANSMIT modes. But every mode needs append data after the i2c address.

@Pillar1989 That means that the original comment is correct and this change is invalid.

> `/* Sending address without data is not supported */`

**Review by stephanosio — CHANGES_REQUESTED** [review]
It looks like the I2C peripheral of the CC13xx and CC26xx SoCs do not support sending address without data as described by the original comment, so this change is invalid.

The current I2C scan implementation expects the I2C peripheral to be able to only send the address to probe the peripherals on the bus, and that is with a good reason because sending data byte can be an intrusive behaviour.

If you want to make the I2C scan function work for this SoC (and others that do not support sending address without data), you will need to modify the I2C scan function itself and introduce something like a "intrusive" scan mode.

cc @nashif since he initially implemented the I2C scan function.

**Comment by Pillar1989** [comment]
https://manpages.debian.org/unstable/i2c-tools/i2cdetect.8.en.html @stephanosio 
I have tried `i2cdetect` on Raspberry Pi，It can't appear your want which just sends the address. It seems that these tools are inherently two-sided.  If you want zephyr to keep the I2C bus uninsurable, I think it should be without introducing `i2c scan`.
image

**Comment by stephanosio** [comment]
The source of the problem is the I2C shell scan command implementation that requires a zero-length message to be sent, which the I2C peripheral of this particular SoC cannot support:

https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/drivers/i2c/i2c_shell.c#L64

The current behaviour of the I2C driver returning `-EIO` when zero-length message is specified is correct because this particular SoC indeed does not support sending a zero-length message.

It does not make sense to change/introduce an incorrect behaviour to the I2C driver in order to support the I2C shell scan command -- this is simply not a correct thing to do.

Instead, the I2C shell scan command should be changed to support the class of devices that do not allow sending zero-length message by design (e.g. through a separate "intrusive scan" feature or by making all scans intrusive; I leave this decision up to whomever maintains this feature).

**Comment by cfriedt** [comment]
> The source of the problem is the I2C shell scan command implementation that requires a zero-length message to be sent, which the I2C peripheral of this particular SoC cannot support:

Has TI stated that? There could be a bug or undersupported feature in the SimpleLink SDK, but I wasn't able to find anything elaborating on the '-ENOTSUP' conclusion in git history.

> Instead, the I2C shell scan command should be changed to support the class of devices that do not allow sending zero-length message by design (e.g. through a separate "intrusive scan" feature or by making all scans intrusive; I leave this decision up to whomever maintains this feature).

I would second that as a feature.

**Comment by cfriedt** [comment]
> > The source of the problem is the I2C shell scan command implementation that requires a zero-length message to be sent, which the I2C peripheral of this particular SoC cannot support:
> 
> Has TI stated that? There could be a bug or undersupported feature in the SimpleLink SDK, but I wasn't able to find anything elaborating on the '-ENOTSUP' conclusion [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_cc13xx_cc26xx.c#L61). It would be surprising if a TI part could not accomplish zero-length transfers, because presumably TI parts have been doing that for a long time.
> 
> @Pillar1989 - have you asked about this on e2e.ti.com?

The same question asked on e2e.

https://e2e.ti.com/support/wireless-connectivity/sub-1-ghz-group/sub-1-ghz/f/sub-1-ghz-forum/898040/cc1310-zero-length-write-over-i2c-to-wake-up-a-sensor

@vaishnav98 - are you aware if anyone at TI has investigated this issue? It seems to be a common problem and has an e2e ticket associated with it.

File: drivers/i2c/i2c_cc13xx_cc26xx.c
Comment (Pillar1989) [reply drivers/i2c/i2c_cc13xx_cc26xx.c:61]: @cfriedt  @stephanosio  Assume that we need to add more code to `i2c  scan` to support some SoC like CC13xx serial. any suggest?  Or maybe this is a half-baked idea, and we still need to fix the problem of sending only addresses.

**Comment by cfriedt** [comment]
I think this would be best handled by pinctrl, to be honest. e.g. a gpio i2c device that shares pins with the dedicated i2c device, add a devicetree property..

**Review by cfriedt — CHANGES_REQUESTED** [review]
Will likely require some work involving the upcoming pinctrl feature

**Pull request closed without merging**
